packages/uma-auth-client/src/hooks/useOAuth.tsx

```diff
@@ -0,0 +1,188 @@
+import { auth } from "@getalby/sdk";
+import { create } from "zustand";
+import { persist } from "zustand/middleware";
+import { fetchDiscoveryDocument } from "./useDiscoveryDocument";
+
+type UmaAuthToken = {
+  access_token: string;
+  refresh_token: string;
+  expires_at: number;
+  nwc_connection_uri: string;
+  commands: string[];
+  budget: string;
+  nwc_expires_at: number;
+};
+
+interface TokenState {
+  accessToken: string;
+  refreshToken: string;
+  expiresAt: number;
+}
+
+interface OAuthState {
+  codeVerifier?: string;
+  token?: TokenState | undefined;
+  identityNpub?: string;
+  identityRelayUrl?: string;
+  redirectUri?: string;
+  nwcConnectionUri?: string;
+  commands?: string[];
+  budget?: string;
+  nwcExpiresAt?: number;
+  setToken: (token?: TokenState) => void;
+  /** The initial OAuth request that starts the OAuth handshake. */
+  initialOAuthRequest: (uma: string) => Promise<void>;
+  /** OAuth token exchange occurs after the initialOAuthRequest has been made. */
+  oAuthTokenExchange: (uma: string) => Promise<void>;
+}
+
+export const useOAuth = create<OAuthState>()(
+  persist(
+    (set, get) => ({
+      identityNpub:
+        "npub1scmpzl2ehnrtnhu289d9rfrwprau9z6ka0pmuhz6czj2ae5rpuhs2l4j9d", // TODO: remove this
+      identityRelayUrl: "wss://nos.lol", // TODO: remove this
+      redirectUri: "http://localhost:3001", // TODO: remove this
+      setToken: (token) =>
+        set({
+          token,
+        }),
+      initialOAuthRequest: async (uma) => {
+        const state = get();
+        const { codeVerifier, authUrl } = await getOAuthUrl(state, uma);
+        set({ codeVerifier });
+        window.location.href = authUrl;
+      },
+      oAuthTokenExchange: async (uma) => {
+        const state = get();
+        const result = await oAuthTokenExchange(state, uma);
+        set(result);
+      },
+    }),
+    {
+      name: "uma-connect",
+      partialize: (state) => ({
+        nwcConnectionUri: state.nwcConnectionUri,
+        codeVerifier: state.codeVerifier,
+      }),
+    },
+  ),
+);
+
+const getOAuthUrl = async (state: OAuthState, uma: string) => {
+  const { identityNpub, identityRelayUrl, redirectUri, codeVerifier } = state;
+  const discoveryDocument = await fetchDiscoveryDocument(uma);
+
+  const authClient = new auth.OAuth2User({
+    client_id: `${identityNpub} ${identityRelayUrl}`,
+    callback: redirectUri!,
+    scopes: [],
+    user_agent: "uma-connect",
+    request_options: {
+      base_url: new URL(discoveryDocument.authorization_endpoint).origin,
```

Comment:
Ah I see why you're asking about base_url now. We actually don't need to pass this at all for generateAuthUrl (you're passing authorizeUrl below), but we do need it for the token requests :-(... see my comment below on that.